ENH: Added more options for formats.style.bar

jmarrec opened the pull request:
- [x]  4 tests added  and passed
 - [x] passes ``git diff upstream/master | flake8 --diff``
 - [x] Update documentation

You can now have the df.style.bar be centered on zero or midpoint value (in
addition to the already existing way of having the min value at the left side of the cell)

----

## Documentation:

I have produced the following example that I think does a pretty good job of explaining the new capabilities

```
import pandas as pd
from IPython.display import HTML

# Case 1: s.min() < 0:
test1 = pd.Series([-100,-60,-30,-20], name='All Negative')
# Case 2: s.max() < 0:
test2 = pd.Series([10,20,50,100], name='All Positive')
# Case 3:
test3 = pd.Series([-10,-5,0,90], name='Both Pos and Neg')

head = """
<table>
    <thead>
        <th>Align</th>
        <th>All Negative</th>
        <th>All Positive</th>
        <th>Both Neg and Pos</th>
    </thead>
    </tbody>

"""

aligns = ['left','zero','mid']
for align in aligns:
    row = "<tr><th>{}</th>".format(align)
    for serie in [test1,test2,test3]:
        s = serie.copy()
        s.name=''
        row += "<td>{}</td>".format(s.to_frame().style.bar(align=align, 
                                                           color=['#d65f5f', '#5fba7d'], 
                                                           width=100).render()) #testn['width']
    row += '</tr>'
    head += row
    
head+= """
</tbody>
</table>"""
        

HTML(head)
```

![screen shot 2016-11-28 at 17 06 21](https://cloud.githubusercontent.com/assets/5479063/20675547/04c4af52-b58d-11e6-9695-48a01c37d522.png)

What do you think?

jmarrec committed:
ENH: Added more options for formats.style.bar

You can now have the bar be centered on zero or midpoint value (in
addition to the already existing way of having the min value at the left side of the cell)

Fixed line too long
`git diff upstream/master | flake8 --diff now passes`

Change the tests to match new float formats.

Added documentation on the new df.style.bar options for align and Colors in the documentation.

Fix versionadded

ENH: Added more options for formats.style.bar

You can now have the bar be centered on zero or midpoint value (in
addition to the already existing way of having the min value at the left side of the cell)

Fixed line too long
`git diff upstream/master | flake8 --diff now passes`

Change the tests to match new float formats.

Added documentation on the new df.style.bar options for align and Colors in the documentation.

Fix versionadded

Check for bad align value and raise. Wrote test for it too

Added a simple example before the parametric one

Added a whatsnew note

Replaced 'self.assertEqual(left, rigth)' by 'assert left == right' like
@TomAugspurger asked

rebased

jmarrec commented on pandas/formats/style.py:
```diff
@@ -987,6 +989,15 @@ def bar(self, subset=None, align='left', axis=0,
+                   " of length 2 [ `color_negative`, `color_positive`] "
+                   "(eg: color=['#d65f5f', '#5fba7d'])\n"
+                   "Only the first two list elements will be used here.")
+            warnings.warn(msg, UserWarning)
```
Should I do `raise TypeError(msg)` instead of issuing a warning?

(removing the line saying that only two elements will be used of course...)

sinhrks replied:
I prefer ``ValueError``

sinhrks commented on pandas/formats/style.py:
```diff
@@ -846,39 +848,166 @@ def set_properties(self, subset=None, **kwargs):
+
+        normed = s * 50 * width / (100 * m)
+
+        base = 'width: 10em; height: 80%;'
```
pls pass ``base`` from ``.bar`` because these are identical.

jmarrec replied:
Done

jreback commented:
can you rebase / update

jorisvandenbossche commented:
cc @TomAugspurger

jmarrec commented:
Sorry for the delay, I'm trying to rebase right now (on 9947a99), but I'm getting a few errors

When doing just 'import pandas as pd':

> AttributeError: module 'pandas' has no attribute 'compat'

When running the 'formats' nosetests:

> AttributeError: module 'pandas.lib' has no attribute 'is_decimal'


Is it only me or is the master broken somehow?

jreback commented:
you need to rebuild the extensions

``python setup.py build_ext --inplace``

generally you should do this anytime you pull from master (and there is an update)

jmarrec commented:
That makes sense! Building now.

jreback commented:
can you rebase / update

jmarrec commented:
Done. When do you think this will be merged in?

jmarrec commented:
The CircleCI has tests that fails, but they run fine on my local branch.

For example:

```
pandas$ nosetests pandas/tests/formats/test_style.py:TestStyler.test_bar_align_mid_all_pos
.
----------------------------------------------------------------------
Ran 1 test in 0.044s

OK
```

jreback commented:
we switched to ``pytest`` recently FYI

take a look here: https://travis-ci.org/pandas-dev/pandas/jobs/206164419

jreback commented on pandas/formats/style.py:
```diff
@@ -851,39 +851,180 @@ def set_properties(self, subset=None, **kwargs):
+    def _bar_left(s, color, width, base):
+        """
+        The minimum value is aligned at the left of the cell
+        .. versionadded:: 0.17.1
```
you don't need the vesionadded tags (and they are outdated)

jmarrec replied:
Are you sure you want me to do that? This method *was* added in 0.17.1, I just modified it to accept a tuple of color.

jmarrec replied:
@jreback told me in an earlier review that I didn't need the versionadded tags for these, so I delete them all. Which?

jreback commented on pandas/tests/formats/test_style.py:
```diff
@@ -348,6 +348,109 @@ def test_bar_0points(self):
         self.assertEqual(result, expected)
 
+    def test_bar_align_zero_pos_and_neg(self):
+        df = pd.DataFrame({'A': [-10, 0, 20, 90]})
```
can you add this issue number as a comment

jmarrec replied:
You mean this pull request number? https://github.com/pandas-dev/pandas/pull/14757?

jreback reviewed (changes requested):
pls add a whatsnew note (and you have some failing tests)

jreback commented:
@TomAugspurger can you review

jreback commented:
@jmarrec I would also take your above picture and add it to the style docs themselves (choose a nice place for it).

jmarrec commented:
My failing tests are in python 2.7 only, because I expect `80.0%` but I get `80%` instead.

Eg:

Result in Python 2.7:

```
 (3, 0): ['width: 10em',
  ' height: 80%',
  'background: linear-gradient(90deg, transparent 0%, transparent 80%, #d65f5f 80%, #d65f5f 100%, transparent 100%)']}
```

expected:
```

 (3, 0): ['width: 10em',
  ' height: 80%',
  'background: linear-gradient(90deg, transparent 0%, transparent 80.0%, #d65f5f 80.0%, #d65f5f 100%, transparent 100%)']}
```

jreback commented:
ok, so can do 1 of 2 things.

- conform both results in py2 and py3 (e.g. you are formatting the string)
or
- use a subtition for that value in the test (with different values for py2 and py3).

ideally first way is better

jreback commented on doc/source/html-styling.ipynb:
```diff
@@ -558,14 +607,16 @@
    "metadata": {},
    "source": [
-    "Use `Styler.set_properties` when the style doesn't actually depend on the values."
+    "New in version 0.19.2 is the ability to customize further the bar chart: You can now have the `df.style.bar` be centered on zero or midpoint value (in addition to the already existing way of having the min value at the left side of the cell), and you can pass a list of `[color_negative, color_positive]`.\n",
```
-> 0.20.0

jreback commented on pandas/formats/style.py:
```diff
@@ -851,39 +851,176 @@ def set_properties(self, subset=None, **kwargs):
             A number between 0 or 100. The largest value will cover ``width``
             percent of the cell's width
+        align : str, default 'left'
+            .. versionadded:: 0.19.2
```
-> 0.20.0
also need a blank line before/after the versionadded tag

jreback commented:
cc @jorisvandenbossche

jorisvandenbossche commented on doc/source/html-styling.ipynb:
```diff
@@ -558,14 +587,16 @@
    "metadata": {},
    "source": [
-    "Use `Styler.set_properties` when the style doesn't actually depend on the values."
+    "New in version 0.20.0 is the ability to customize further the bar chart: You can now have the `df.style.bar` be centered on zero or midpoint value (in addition to the already existing way of having the min value at the left side of the cell), and you can pass a list of `[color_negative, color_positive]`.\n",
```
The table below is a nice overview. But it is also a bit hard to interpret as a user what is the actual code he/she should use. Can you add first an example with the existing frame? Eg `df.style.bar(subset=['B'], align='mid')`. And then give the overview table.

jorisvandenbossche commented on pandas/formats/style.py:
```diff
@@ -851,39 +851,178 @@ def set_properties(self, subset=None, **kwargs):
             a valid slice for ``data`` to limit the style application to
         axis: int
-        color: str
+        color: str (for align='left') or 2-tuple/list (for align='zero', 'mid')
```
In the `_bar_left` function also a tuple is accepted, which contradicts with this

jmarrec replied:
Thanks for catching that.

jorisvandenbossche commented on pandas/formats/style.py:
```diff
@@ -851,39 +851,178 @@ def set_properties(self, subset=None, **kwargs):
+            - 'zero' : a value of zero is located at the center of the cell
+            - 'mid' : the center of the cell is at (max-min)/2, or
+                if values are all negative (positive) the zero is aligned
+                at the right (left) of the cell
```
I find the names a bit confusing. As 'zero' is actually more 'centered'? While I first thought 'mid' would give me centered bars, but that depends on the actual values you have.

BTW, 'mid' actually seems a more sensible default. Maybe we should change that? (cc @TomAugspurger)

TomAugspurger replied:
> I find the names a bit confusing. As 'zero' is actually more 'centered'?

I think this is only true if you have all positive values.  So

```python
pd.DataFrame(np.arange(-2, 5)).style.bar(align='zero')
```

Makes sense to me since zero is at the middle.

> BTW, 'mid' actually seems a more sensible default.

Probably, but I'm not sure it's worth breaking API. Maybe with a deprecation cycle?

jmarrec replied:
@jorisvandenbossche: @TomAugspurger is correct, it's really centered on zero. The example I tried to include in the doc and that I pasted in the first post of this PR should highlight that fact, otherwise I should revise it.

I have no idea how you handle deprecation cycles etc, so I'll leave that completely up to you guys.

jorisvandenbossche commented on pandas/formats/style.py:
```diff
@@ -851,39 +851,178 @@ def set_properties(self, subset=None, **kwargs):
+                       color=color, width=width, base=base)
+        elif align == 'mid':
+            self.apply(self._bar_center_mid, subset=subset, axis=axis,
+                       color=color, width=width, base=base)
```
Can you do here an "else" clause to raise a ValueError when a not accepted value for align was passed? and also add a test for it. 
(with playing a bit, I already was bitten by this by trying to pass 'right' and 'center' to align, and thinking that it was not working :-))

jorisvandenbossche reviewed (commented):
Nice enhancement! Added some comments

jmarrec commented:
@jorisvandenbossche I haven't forgot about your comments, it just fell on my back burner. I'll try to get to it asap in the coming days.

jreback commented:
@TomAugspurger status?

TomAugspurger reviewed (changes requested):
This is really close. I think just adding the test for raising with invalid `align`s. Do you have time to update @jmarrec ?

jmarrec commented:
@TomAugspurger Really sorry, it had fallen on the back burner. I'll get on it this week, will try for tomorrow, so I can put this to bed.

jmarrec commented:
Yikes, this looks really messy sorry. For some reason when I rebased on upstream/master today, my branch diverged. I merged origin/style-bar back into my local branch, and pushed... I didn't know it'd clutter this PR that much, and now that I pushed I don't want to rewrite history.

Anyways, I think I made all the changes you three requested, except the versionadded tags on each of the `_bar_xxx` methods since @jreback had told me to delete them and it seems @TomAugspurger reviewed an older version where they were still there but outdated (I have deleted them a while ago). If you want me to add them I guess I can. Otherwise I hope this is ready to go!

jreback commented:
@jmarrec looks like you pushed all commits, do this

```
git rebase -i thisbranch origin/master
git push yourorigin thisbranch -f
```

to fix

jmarrec commented:
@jreback in your example 'origin' means this 'pandas-dev' repo right?

I end up with a detached HEAD, I'm afraid of force pushing. Is that expected?

jreback commented:
i'll fix your branch give me about 15 min

jmarrec commented:
@jreback will this make it in (or through #15900, now closed)?

jreback commented:
see the comments - iirc needs some more tests

jmarrec commented:
Well, missing I'm missing something obvious, but I don't see any comments left to address, that's precisely why I'm asking.

Like I wrote a month ago, the only "open" things is what @TomAugspurger asked, that is to change the version added tags, but he was reviewing an outdated version, because before that you already told me to delete these version added tags, which I did.

TomAugspurger commented:
I think @jorisvandenbossche's comment https://github.com/pandas-dev/pandas/pull/14757/files/b22f6392b3416cdc02138a79efb8f179751c8141#r104129905 is still outstanding. Handling cases like `.bar(align='foo')`. That should raise with something like `ValueError('align' must be one of {'left', 'zero', 'mid'}, got 'foo' instead)`

Also can you change all of the `self.assertEqual(left, right)`s to `assert left == right`?

jmarrec commented:
Was done on April 5, see https://github.com/jmarrec/pandas/blob/style-bar/pandas/formats/style.py#L1034.

I did the change you requested (in all the file not just my changes, used a regex...).

Now, is everything ok?

TomAugspurger commented:
Whoops, looks like some git issues. Let me see if I can rebase for you, one sec.

jmarrec commented:
If you find how to fix it (Jreback did earlier), if you could please share which commands did the trick, that'd be great thanks

TomAugspurger commented:
Typically it's just `git fetch upstream`, then `git rebase upstream/master`. This was a bit tricker than usual because it's been sitting out for so long (sorry about that) and we've re-arranged a lot of things now.  This causes conflicts, since two commits touched the same lines so you have to go in and fix them manually with your editor. Then it's `git add <files>` and `git rebase --continue.` until you're fully rebased.

https://github.com/pandas-dev/pandas/pull/14757/commits/ca9c0d546833e25d08f080059e71e89bb153d7a2 *should* have everything down in a single commit, if you want to give that a skim to make sure I didn't miss anything.

jorisvandenbossche committed:
small doc fixes

jorisvandenbossche reviewed (approved):
Added a commit with a small doc correction (indentation and white space). And also aligned the signature with documentation, adding the new keyword at the end, not breaking people who used axis positional

jorisvandenbossche commented:
Let's merge this PR, but, I find we should think about changing the default value. If we all think 'mid' is the more sensible default, IMO we should just change it. 
It's a while that I look into the PR in detail, but if I recall correctly, the 'mid' argument would only change things in case of negative values? (which I think is a less occurring usecase, and currently also a bid odd result?) If that is correct, I would opt to just change it.

gfyoung commented:
@TomAugspurger : For subsequent PR's, could you check to make sure that old `nosetest` idiom is not being used?  I found a couple in the diff that you just merged in.

TomAugspurger commented:
My apologies. I thought I caught all those in the rebase :/

I'll submit a fix now.

gfyoung commented:
No worries.  I'm putting up a PR to remove vestigial statements like these, so that won't be necessary.

I just wanted to alert you about it for subsequent PR's.

halflings commented:
This is breaking the previous behavior of Pandas when using `.plot.bar` : passing a list of colors (with 2 elements or more) would give each bar a different color, as it is the case in `matplotlib`.
I also think passing an array with two values which are implicitly interpreted as [negative_color, positive_color] is confusing, and we should have instead two explicit parameters for this feature.

Example of bar coloring breaking in 0.20:

    import pandas as pd
    df = pd.DataFrame(dict(created=[0, 1, 2, 3], amount=[6, 1, 2, 5], is_active=[0, 1, 0, 1]))
    df.set_index('created').plot.bar(y='amount', color=df.is_active.apply(lambda v : ['red', 'blue'][v]

Shows this in previous verisons:

![download](https://user-images.githubusercontent.com/1626936/27826397-bcb6d7ac-60aa-11e7-8a06-b3978c6da623.png)
))

Everything is blue in 0.20+
